### DISCUSSION TITLE
Using Jupyterlite with git-wasm

### DISCUSSION THREAD

**Discussion by philipp-strack** [jupyterlite/jupyterlite#740] [Ideas]
Hello,

I want to integrate jupyterlite with wasm-git

https://github.com/petersalomonsen/wasm-git

to enable pulling a git repository to the in browser memory filsystem and pushing changes to a git server.

Do you think this is feasible? How would you integrate wasm-git with jupyterlite?

Thanks a lot!

**Comment by bollwyvl** [comment]
This would be very cool, and answer a lot of use cases.

It is feasible, but there are a number of ways of actually mechanically doing it.

You could likely build a no-JS prototype of this with:
- a notebook
- the pyolite kernel
- a noarch build of [dulwich](https://github.com/jelmer/dulwich)
- [ipylab](https://github.com/jtpio/ipylab) for ui

While not shippable, it would potentially reveal some limitations.

I think the cleanest way to do it would be emulate the `IDrive` [jupyterlab-filesystem-access](https://github.com/jupyterlab-contrib/jupyterlab-filesystem-access) that works in both full JupyterLab and JupyterLite, as this would be useful _anyways_ for folks that don't have a "real" git. The `IDrive` stuff means you get a separate file browser UI, extended with some inputs/buttons that lets you pick a remote, clone it, and start working with stuff, then a simplified "commit and push" button. Otherwise everything works the same, and of late, would appear inside the `/drive` folder iniside running WASM kernels.

Another approach might be to implement the backend of [jupyterlab-git](https://github.com/jupyterlab/jupyterlab-git), though this might be rather involved.

At some point in the future, we might adopt the emscripten filesystem as _the_ basic backend, but that is partially blocked by [WasmFS](https://emscripten.org/docs/api_reference/Filesystem-API.html#new-file-system-wasmfs), as that would allow us to simplify much of our handling of contents.

**Comment by philipp-strack** [comment]
Thanks a lot for your quick reply, that is super helpful!

I am new to the project so please excuse the naive question: Is it possible to mount the emscripten filesystem created by wasm-git into the filesystem tree created by jupiterlite?
